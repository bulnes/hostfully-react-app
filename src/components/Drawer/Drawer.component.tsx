import { useContext } from "react";
import { BookingsContext } from "../../@contexts/BookingsContext";
import { BookingForm } from "../BookingForm/BookingForm.component";

export function Drawer() {
  const { showBookings, setShowBookingsVisibility, bookings } =
    useContext(BookingsContext);

  return (
    <div className={`${showBookings ? "block" : "hidden"}`}>
      <div className="bg-gray-900/50 fixed inset-0 z-30"></div>

      <div className="fixed top-0 right-0 z-40 h-screen p-4 overflow-y-auto transition-transform bg-white w-80 lg:w-96">
        <h5 className="inline-flex items-center mb-4 text-base font-semibold text-gray-500 dark:text-gray-400">
          <svg
            className="w-4 h-4 mr-2"
            aria-hidden="true"
            xmlns="http://www.w3.org/2000/svg"
            fill="currentColor"
            viewBox="0 0 20 20"
          >
            <path d="M10 .5a9.5 9.5 0 1 0 9.5 9.5A9.51 9.51 0 0 0 10 .5ZM9.5 4a1.5 1.5 0 1 1 0 3 1.5 1.5 0 0 1 0-3ZM12 15H8a1 1 0 0 1 0-2h1v-3H8a1 1 0 0 1 0-2h2a1 1 0 0 1 1 1v4h1a1 1 0 0 1 0 2Z"></path>
          </svg>
          Bookings
        </h5>

        <button
          type="button"
          className="text-gray-400 bg-transparent hover:bg-gray-200 hover:text-gray-900 rounded-lg text-sm w-8 h-8 absolute top-3 end-3 flex items-center justify-center"
          onClick={() => setShowBookingsVisibility(false)}
        >
          <svg
            className="w-3 h-3"
            aria-hidden="true"
            xmlns="http://www.w3.org/2000/svg"
            fill="none"
            viewBox="0 0 14 14"
          >
            <path
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              d="m1 1 6 6m0 0 6 6M7 7l6-6M7 7l-6 6"
            ></path>
          </svg>
          <span className="sr-only">Close menu</span>
        </button>

        <div className="mt-8">
          {bookings.length > 0 ? (
            <ul>
              {bookings.map((booking) => (
                <li key={booking.id} className="mb-8">
                  <div className="flex items-start justify-between flex-col mb-2">
                    <h6 className="text-lg font-semibold">{booking.title}</h6>
                    <span className="text-gray-500 dark:text-gray-400">
                      ${booking.pricePerNight} / night
                    </span>
                  </div>

                  {booking.checkInDate && booking.checkOutDate && (
                    <p className="text-gray-500 dark:text-gray-400">
                      {booking.checkInDate} - {booking.checkOutDate}
                    </p>
                  )}

                  <BookingForm formType="update" {...booking} />
                </li>
              ))}
            </ul>
          ) : (
            <p className="text-gray-500 dark:text-gray-400">
              No bookings yet. Start booking now!
            </p>
          )}
        </div>
      </div>
    </div>
  );
}
