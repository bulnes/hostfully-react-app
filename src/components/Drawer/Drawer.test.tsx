import { render } from "@testing-library/react";
import { Drawer } from "./Drawer.component";

const setShowBookingsVisibility = jest.fn();

// Mocking the useContext hook
jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useContext: () => ({
    showBookings: true,
    setShowBookingsVisibility,
    bookings: [
      {
        id: 1,
        title: "Sample booking",
        pricePerNight: 100,
      },
    ],
  }),
}));

describe("Component: Drawer", () => {
  it("should be defined", () => {
    expect(Drawer).toBeDefined();
  });

  it("should render correctly", () => {
    const wrapper = render(<Drawer />);
    expect(wrapper).toBeDefined();
  });

  it("should render bookings correctly", () => {
    const wrapper = render(<Drawer />);
    expect(wrapper.getByText("Bookings")).toBeDefined();
  });

  it("should render close button correctly", () => {
    const wrapper = render(<Drawer />);
    expect(wrapper.getByText("Close menu")).toBeDefined();
  });

  it("should render booking title correctly", () => {
    const wrapper = render(<Drawer />);
    expect(wrapper.getByText("Sample booking")).toBeDefined();
  });

  it("should trigger setShowBookingsVisibility on close button click", () => {
    const wrapper = render(<Drawer />);
    wrapper.getByText("Close menu").click();
    expect(setShowBookingsVisibility).toHaveBeenCalled();
  });

  it("should be represented by the snapshot", () => {
    const { container } = render(<Drawer />);
    expect(container).toMatchSnapshot();
  });
});
