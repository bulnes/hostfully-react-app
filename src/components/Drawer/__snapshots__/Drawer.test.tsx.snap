// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Drawer should be represented by the snapshot 1`] = `
<div>
  <div
    class="block"
  >
    <div
      class="bg-gray-900/50 fixed inset-0 z-30"
    />
    <div
      class="fixed top-0 right-0 z-40 h-screen p-4 overflow-y-auto transition-transform bg-white w-80 lg:w-96"
    >
      <h5
        class="inline-flex items-center mb-4 text-base font-semibold text-gray-500 dark:text-gray-400"
      >
        <svg
          aria-hidden="true"
          class="w-4 h-4 mr-2"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10 .5a9.5 9.5 0 1 0 9.5 9.5A9.51 9.51 0 0 0 10 .5ZM9.5 4a1.5 1.5 0 1 1 0 3 1.5 1.5 0 0 1 0-3ZM12 15H8a1 1 0 0 1 0-2h1v-3H8a1 1 0 0 1 0-2h2a1 1 0 0 1 1 1v4h1a1 1 0 0 1 0 2Z"
          />
        </svg>
        Bookings
      </h5>
      <button
        class="text-gray-400 bg-transparent hover:bg-gray-200 hover:text-gray-900 rounded-lg text-sm w-8 h-8 absolute top-3 end-3 flex items-center justify-center"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="w-3 h-3"
          fill="none"
          viewBox="0 0 14 14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m1 1 6 6m0 0 6 6M7 7l6-6M7 7l-6 6"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
        <span
          class="sr-only"
        >
          Close menu
        </span>
      </button>
      <div
        class="mt-8"
      >
        <ul>
          <li
            class="mb-8"
          >
            <div
              class="flex items-start justify-between flex-col mb-2"
            >
              <h6
                class="text-lg font-semibold"
              >
                Sample booking
              </h6>
              <span
                class="text-gray-500 dark:text-gray-400"
              >
                $
                100
                 / night
              </span>
            </div>
            <form>
              <div
                class="flex items-center gap-3 mt-3 mb-5"
              >
                <div
                  class="flex items-start flex-col gap-1"
                >
                  <label
                    class="text-gray-600"
                    for="checkin"
                  >
                    Check-in
                  </label>
                  <input
                    class="border rounded-md p-1"
                    id="checkin"
                    type="date"
                    value=""
                  />
                </div>
                <div
                  class="flex items-start flex-col gap-1"
                >
                  <label
                    class="text-gray-600"
                    for="checkout"
                  >
                    Check-out
                  </label>
                  <input
                    class="border rounded-md p-1 w-full"
                    id="checkout"
                    type="date"
                    value=""
                  />
                </div>
              </div>
              <div
                class="flex items-center gap-3 mt-4"
              >
                <button
                  class="bg-white text-blue-700 border border-blue-500 font-bold py-2 px-4 rounded"
                  type="button"
                >
                  Update
                </button>
                <button
                  class="bg-white text-red-700 border border-red-500 font-bold py-2 px-4 rounded"
                  type="button"
                >
                  Delete
                </button>
              </div>
            </form>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
